Adding shader preprocessor support

TheOrangeDay opened the pull request:
Reviving #17797

This takes the work from @bitsawer above and gets it up-to-date in master. I've also taken some code from #41583 To use the shader dependency graph implementation from @MagdielM

The shader preprocessing supports:
if, ifdef, ifndef, else, endif, define, undef, and include

It does not support the operation "defined(...)" yet.
The "if" parser uses the Expression implementation to evaluate the conditions.

At the moment the define method can define _constant_ values (integers, floats, strings, etc) -- defining a method is not supported yet.

One thing I'm not clear about is the mention by @reduz of supporting constness (tbh not sure what this refers to). It appears from @aaronfranke that this has been handled now?

With this code in place, you can write shaders like:
`uber.gdshader`
```
shader_type spatial;

#ifdef USE_COLOR
	instance uniform vec3 object_color;
#endif

#ifdef USE_OBJECT_ALPHA
	instance uniform float object_alpha = 1.0f;
#endif

void fragment()
{
#ifdef USE_COLOR
	ALBEDO = object_color;
#endif

#ifdef USE_OBJECT_ALPHA
	ALPHA = object_alpha;
#endif
}`

and then have a specific gdshader that can enable defines:
colored_alpha.gdshader
`shader_type spatial;

#define USE_COLOR
#define USE_OBJECT_ALPHA

#include "uber.gdshader"
```

You could also use it to write a library of useful methods and just #include it as needed. Perhaps it would be possible to use a library like the following as well:
https://github.com/patriciogonzalezvivo/lygia


Please take a look at the code and let me know if something needs to be adjusted. This will be my first contribution to the project, so I'd like to make sure everything looks right.

<i>Bugsquad edit</i>: Closes https://github.com/godotengine/godot-proposals/issues/944

TheOrangeDay committed:
Adding support for shader preprocessing

This revives the preprocessor code from bitsawer and gets it up-to-date with master. I've switched the conditional parsing to use the Expression implementation. I also added support for #ifndef

This also includes code from MagdielM to use his ShaderDependencyGraph code. I made changes to avoid updating previously visited included shaders and switched to use the tokenizer to look for #include uses

Co-Authored-By: bitsawer <22456603+bitsawer@users.noreply.github.com>
Co-Authored-By: Magdiel <56076033+MagdielM@users.noreply.github.com>

TheOrangeDay committed:
Updating preprocessor comment stripping to ignore invalid characters and updated the shader dependency parser to be safer when code is pulled during mid-authoring

aaronfranke commented:
> One thing I'm not clear about is the mention by @reduz of supporting constness (tbh not sure what this refers to). It appears from @aaronfranke that this has been handled now?

I have no idea what you're referring to, sorry.

fire commented:
#define and if else 

I'll need to ask the other maintainers.

When I last investigated there was duplicate functionality in `if, ifdef, ifndef, else, endif, define, undef, and include`.

So I tried to determine what doesn't exist in Godot Engine.

1. `if else` statements exist
2. `define`s are similar to uniforms
3. `include` is missing

I'll try to get some commentary from some of the rendering team @reduz @clayjohn @Chaosus

briansemrau commented:
@fire, I wouldn't consider any of these preprocessor statements to be redundant. While you can use things like uniforms and if/else code to produce the same behavior, these have an impact on performance at runtime. Excessive branching can be costly and uniforms are not unlimited. Having _all_ of these features would be a significant usability boost for the godot shading language.

Chaosus commented:
I have nothing against this feature if this is properly implemented, it is a definitely huge usability boost and it would really speed up porting shaders from ShaderToy. As far as I remember the main concern for @reduz was slowing down the shader compilation speed.

PS: I see this PR not touching the shader compiler to push preprocessor directives to it - that's cool.

fire commented:
I am not against this feature. Please count me as abstaining from consensus.

Ansraer commented:
Oh, this would be great for plugin authors!
I noticed that currently the user has to manually set defines in the shader code. It would be great if that part could be somehow exposed to the inspector, so that users could define/undefine stuff simply by ticking a checkbox in the inspector.

Another thing I would like to see is a new shader type "include" that can't be instanced for utility code that is not supposed to work on its own.

TheOrangeDay commented:
> > One thing I'm not clear about is the mention by @reduz of supporting constness (tbh not sure what this refers to). It appears from @aaronfranke that this has been handled now?
> 
> I have no idea what you're referring to, sorry.

Hey @aaronfranke, sorry about that. I was mistaken, it was @akien-mga that commented about work @Chaosus was doing to support constness in comment: 
https://github.com/godotengine/godot/pull/17797#issuecomment-525665029\
Sounds like that's been addressed.

> Oh, this would be great for plugin authors! I noticed that currently the user has to manually set defines in the shader code. It would be great if that part could be somehow exposed to the inspector, so that users could define/undefine stuff simply by ticking a checkbox in the inspector.
> 
> Another thing I would like to see is a new shader type "include" that can't be instanced for utility code that is not supposed to work on its own.

@Ansraer I like the idea of having editable #define's external of the shader. I wonder if it would be worth having those set up in the project settings as default defines for all shader builds like you'd find in code compiler settings for example. In the case of support for toggling and creating defines within the material... that would take more consideration, especially with parsing and pushing into the UI.

As for a shader_type that's for includes only -- I can look into something like "library" as the value for that. I'm not familiar enough with the validation feedback from the material:shader link to know where I would go about throwing an error for incorrect usage, but I can poke around and see if I can find a place for that.


@Chaosus If there was a performance concern from @reduz, I'd be happy to collect some metrics from a larger project to make sure I haven't missed anything that could slow down existing workflows significantly. Anyone know of a project that might be a good candidate I can open up and run some tests on?

Ansraer commented:
As someone who just started working on their first shader addon I would really appreciate it if I could expose defines to the materials UI. Currently I have tons of if statements and always ~50-70% unused code.
I am not 100% sure how to best implement this though. At least the part of godot I have looked at recently compiles all possible permutations of defines and picks the correct pipeline at runtime. However, given how fast the number of permutations increases when you add additional defines I doubt that this would scale well. Perhaps you could somehow only compile the permutations that are actually requested by materials and cache the result so identically setup materials can use it?

Global defines are another great idea I hadn't even considered. We could even have some default #defines for stuff like e.g. the used renderer. Would allow shaders to dynamically adopt to the targeted hardware.
Could probably also be used to create optional dependencies on other addons if they set their own define when they are added to a project.

clayjohn commented:
> Oh, this would be great for plugin authors! I noticed that currently the user has to manually set defines in the shader code. It would be great if that part could be somehow exposed to the inspector, so that users could define/undefine stuff simply by ticking a checkbox in the inspector.

We currently have a ``custom defines`` property that lets you add in defines to shaders https://docs.godotengine.org/en/3.4/classes/class_shader.html#class-shader-property-custom-defines It was added to allow for enabling extensions for VR applications, but I think it would work well here. 

With respect to this PR. One of the significant challenges other attempts have faced is proper error reporting and validation. For example, if there is a clash between definitions in the included files, how does the editor draw the users attention to the right line in the right file to resolve the error?

fire commented:
All the tests are failing and need to resolved to pass to the next stage. Also, some design comments were mentioned.

TheOrangeDay commented:
@clayjohn I've been thinking about how to better guide users to the line that's causing the error. I think to properly support this we need to have a more up front display of shader dependencies within the editor dialog and allow viewing the dependent code within the context of the compiling shader. I'll look at what we can do to better track the relationships and the locations where the errors come up and push a more user friendly update.

@fire I've fixed the issue causing the failed checks, but I'm going to wait before updating the PR until I can figure out a better way to display the errors.

TheOrangeDay committed:
Adding support for shader dependency graph as a tree within the shader editor directly

This primes the editor for better editing and error contexts. This can also be expanded to enhance the highlighting of code because you'll see the code that's included in full dependency context based on available defines (that work is pending)

Also fixing forward declare issue (used class instead of struct, this was causing a build error with "warnings as errors")

TheOrangeDay committed:
Adding support for shader dependency graph as a tree within the shader editor directly

This primes the editor for better editing and error contexts. This can also be expanded to enhance the highlighting of code because you'll see the code that's included in full dependency context based on available defines (that work is pending)

Also fixing forward declare issue (used class instead of struct, this was causing a build error with "warnings as errors")

TheOrangeDay commented:
Didn't realize that new commits were automatically added to the PR. As of right now, this PR is WIP.
Sorry if the new commits spam anyone. I come from a different version control system (p4), so still getting familiar with git.

TheOrangeDay committed:
WIP - added more contextual shader debugging. Brings support for resolving error line back to source file and opening in the editor to highlight issues. Tried switching to FileAccess instead of ResourceLoader, but going to switch back because this won't work for resolving binary data in finalized builds.

TheOrangeDay committed:
Primitive support for indicating in the editor which #if blocks are going to be compiled

This updates the text editor implementation to carry a color_scale value per line. May change to font_color_intensity as it's probably clearer that way

TheOrangeDay committed:
renamed color_scale to font_color_intensity + starting to fix contextual error location

TheOrangeDay committed:
Expanding support for goto error, fixing issues with shader context resolve, updated editor minimap to use font intensity, added error background color to shader dep tree for errored shader includes

TheOrangeDay committed:
Addressing some of the static check issues

TheOrangeDay committed:
reorganizing shader dependency panel

TheOrangeDay committed:
fixing compilation

TheOrangeDay committed:
More formatting updates

TheOrangeDay committed:
Updating some files with clang-format

TheOrangeDay committed:
More changes as a result of clang-format

TheOrangeDay committed:
Last of the clang-format checks

TheOrangeDay committed:
Switching back to resource loader

Chaosus commented:
@TheOrangeDay The build errors should be fixed (just a reminder).

TheOrangeDay commented:
@Chaosus Thanks for the reminder. I've been working on the shader preprocessor in a WIP branch in my fork, here:
https://github.com/TheOrangeDay/godot/tree/shader-preprocessor-wip

Where I've been addressing some usability concerns. I was thinking about how to set up the editing context for a given shader to better communicate errors to the user, and I've went in the direction of using some of the preprocessing within the shader editor to find references and then hold on to the defines and evaluated conditions.

![image](https://user-images.githubusercontent.com/6472143/142176405-7993499a-9784-4ffb-a12f-23c07dd983af.png)

With all the includes and conditions evaluated ahead of time, we can use this information to better communicate a full editing context. I've added a tree view to the right of the shader editor to show your includes and updated the shader editor text editor to darken code that's excluded by failing conditions (see #if PI < test in above screenshot).

I've also updated the error reporting so that when compiling the main shader (alpha.gdshader in this case), that if there are failures in child included shaders, the file with the errors is highlighted in the tree view. When you click the error at the bottom panel, or double click on the highlighted tree item, it opens that file in the same editing context (maintaining the defined macros), and highlights the errored line.

![image](https://user-images.githubusercontent.com/6472143/142177162-11fd2a0e-992a-4a12-8c24-0f52af081b77.png)

There are some additional changes to be done to round off the work (eg path name in error just to reinforce location) -- the code is kind of in flux at this point. 

From what I've described above, does this sound okay? Do you think this addresses usability concerns? I haven't evaluated performance yet as I want to get all the features required to make this a viable change first, then see what can be done to maintain the features without much additional overhead.

Chaosus commented:
I think sounds ok (I would test it a bit when I have free time). That right panel could be hidden and is hidden by default? I would also restrict using built-in names for defines (such as PI) or emit warnings at least. Don't bother much about performance - it was just an old @reduz concern.

TheOrangeDay committed:
Fixing error line output from shader rd, adding path to error output in shader editor, fixing contextual validation

TheOrangeDay commented:
> I think sounds ok (I would test it a bit when I have free time). That right panel could be hidden and is hidden by default? I would also restrict using built-in names for defines (such as PI) or emit warnings at least. Don't worry much about performance - it was just an old @reduz concern - maybe he thinks another way now.

If you could test, that would be great. Right now the implementation is pretty functional, but there are some more features that need to go in. After that, I will be doing a refactoring sweep.

Good call on hiding the dependency tree. I can add a new toolbar menu, something like "View -> Shader Dependencies"

Just to note which features are required that I'll be looking into next:
- Update shader dependency tree when code changes
- Ensure errors from preprocessor failing can be properly resolved to file that caused the issues
- Update shader compiler to accept shader context to prevent continual compile of child shaders
  - I think this comes as a result of the switch to the use of the resource loader. Doesn't hurt to be able to pass in a shader context anyway. I generate a preprocessed context way too often right now.
- Evaluate packaged build system to collapse shaders with includes to prevent preprocessing in final game builds.
  - If we do this, we could switch back to FileAccess instead of ResourceLoader, but lose the ability to reference .tres files.

Chaosus commented on editor/plugins/shader_editor_plugin.cpp:
```diff
@@ -54,14 +55,18 @@ Ref<Shader> ShaderTextEditor::get_edited_shader() const {
+	set_edited_shader(p_shader, p_shader->get_code());
+}
+
+void ShaderTextEditor::set_edited_shader(const Ref<Shader>& p_shader, String code) {
```
String code -> String p_code

Chaosus commented on editor/plugins/shader_editor_plugin.cpp:
```diff
@@ -94,6 +99,10 @@ void ShaderTextEditor::set_warnings_panel(RichTextLabel *p_warnings_panel) {
 	warnings_panel = p_warnings_panel;
 }
 
+void ShaderTextEditor::set_shader_dependency_tree(Tree* tree) {
```
Tree* tree -> Tree* p_tree

Chaosus commented on editor/plugins/shader_editor_plugin.cpp:
```diff
@@ -649,6 +694,80 @@ void ShaderEditor::_make_context_menu(bool p_selection, Vector2 p_position) {
+	}
+}
+
+void ShaderEditor::_update_shader_dependency_tree_items(TreeItem* parent_tree_item, ShaderDependencyNode* node)
```
Add `p_` before parameter names.

Chaosus commented on editor/plugins/shader_editor_plugin.h:
```diff
@@ -132,6 +145,10 @@ class ShaderEditor : public PanelContainer {
 
+	void _tree_activate_shader();
+	void _update_shader_dependency_tree();
+	void _update_shader_dependency_tree_items(TreeItem* parent_tree_item, ShaderDependencyNode* node);
```
Add `_p` before parameter names.

Chaosus commented on editor/plugins/shader_editor_plugin.h:
```diff
@@ -71,8 +75,13 @@ class ShaderTextEditor : public CodeTextEditor {
 	void set_warnings_panel(RichTextLabel *p_warnings_panel);
 
+	// TODO even necessary now?
+	void set_shader_dependency_tree(Tree* tree);
```
Add `_p` before the parameter name.

Chaosus commented on editor/plugins/shader_editor_plugin.h:
```diff
@@ -53,11 +55,13 @@ class ShaderTextEditor : public CodeTextEditor {
 
 	void _check_shader_mode();
 	void _update_warning_panel();
+	void _update_shader_dependency_tree_items(TreeItem* parent_tree_item, ShaderDependencyNode* node);
```
Add `_p` before parameter names.

Chaosus commented on servers/rendering/shader_preprocessor.h:
```diff
@@ -0,0 +1,144 @@
+	List<ShaderDependencyNode*> cyclic_dep_tracker;
+	List<Ref<Shader>> visited_shaders;
+
+	Set<ShaderDependencyNode*>::Element* find(Ref<Shader>);
```
Add `p_shader` parameter name.

Chaosus commented on servers/rendering/shader_preprocessor.h:
```diff
@@ -0,0 +1,144 @@
+	List<Ref<Shader>> visited_shaders;
+
+	Set<ShaderDependencyNode*>::Element* find(Ref<Shader>);
+	void populate(ShaderDependencyNode*);
```
Add `p_node` parameter name.

Chaosus commented on servers/rendering/shader_preprocessor.h:
```diff
@@ -0,0 +1,144 @@
+
+	Set<ShaderDependencyNode*>::Element* find(Ref<Shader>);
+	void populate(ShaderDependencyNode*);
+	void update_shaders(ShaderDependencyNode*);
```
Add `p_node` parameter name too.

Chaosus commented on servers/rendering/shader_preprocessor.cpp:
```diff
@@ -0,0 +1,955 @@
+	}
+}
+
+void ShaderDependencyGraph::update_shaders(ShaderDependencyNode* node) {
```
node -> p_node

Chaosus commented on servers/rendering/shader_preprocessor.cpp:
```diff
@@ -0,0 +1,955 @@
+	cyclic_dep_tracker.erase(node);
+}
+
+Set<ShaderDependencyNode*>::Element* ShaderDependencyGraph::find(Ref<Shader> shader) {
```
shader -> p_shader

Chaosus commented on servers/rendering/shader_preprocessor.cpp:
```diff
@@ -0,0 +1,955 @@
+	populate(node);
+}
+
+void ShaderDependencyGraph::populate(ShaderDependencyNode* node) {
```
node -> p_node

Chaosus commented:
@TheOrangeDay  You need to add `p_` before all parameter names, (add `r_` if they are used as reference), remove all unnamed parameters.

Regarding TheOrangeDay/godot@shader-preprocessor-wip - please rebase it to the newest version of the engine (for example 4ea87f36cd23494bfbde800b436467d3d0f82b52) - since I'm a bit lazy to recompile engine to the older versions.

TheOrangeDay committed:
Replacing erase with substr as erase is deprecated now.

TheOrangeDay committed:
Updating dependencies and tree on change, fixing line context resolve, updating warnings to use the line context resolve

TheOrangeDay committed:
Updating dependencies and tree on change, fixing line context resolve, updating warnings to use the line context resolve. cont'd

TheOrangeDay committed:
Updating parameters to have p_ and r_, reorganized initializers in Shader Dependency Nodes

Chaosus commented:
@TheOrangeDay - I finally tested it on the updated branch! That's magnificent work and works as expected.

PS: I think printing error("full path to file":line) should be prevented if only one file is used. Leave it for the shaders with at least one #include.

TheOrangeDay commented:
> @TheOrangeDay - I finally tested it on the updated branch! That's magnificent work and works as expected.
> 
> PS: I think printing `error("full path to file":line)` should be prevented if only one file is used (print `error(line):` instead). Leave it for the shaders with at least one `#include`.

Hey @Chaosus, thanks for the review, I really appreciate it! I'm glad you like the update!

I'm going to finish off the remaining features I mentioned earlier, and then merge over my wip branch into this request:
>     * Ensure errors from preprocessor failing can be properly resolved to file that caused the issues
> 
>     * Update shader compiler to accept shader context to prevent continual compile of child shaders
>       
>       * I think this comes as a result of the switch to the use of the resource loader. Doesn't hurt to be able to pass in a shader context anyway. I generate a preprocessed context way too often right now.
> 
>     * Evaluate packaged build system to collapse shaders with includes to prevent preprocessing in final game builds.
>       
>       * If we do this, we could switch back to FileAccess instead of ResourceLoader, but lose the ability to reference .tres files.


Taking a look at the compile checks from my recent submit, looks like I also need to refactor out the use of Regex in the preprocessor as well since it's only allowed in the editor. However if I can get "post processed" shaders working in the packaged builds, I don't think I need to remove the regex implementation.

One more thing to note about this to make sure it's clear what this preprocessor implementation is doing -- the preprocessor macros are not passed through to the shader compiler, it's all processed out ahead of time during authoring. This is so we can parse, evaluate, and report errors as they come up to the user. If a user has defined some custom defines on the shader object, those are only passed through at final compilation of the shader. At this point i have not pulled the custom defines for use in preprocessing.

Chaosus commented:
I would also add a directive `#pragma disable_preprocessor` to provide a way to speed up a compilation of some shaders which does not use a preprocessor.

bitsawer commented:
Great to see this getting some attention again, I'm the original author of that PR #17797. Looks like @TheOrangeDay has done a lot of good extra work to improve the functionality. Feel free to use and modify it as much as is needed. Some possibly useful notes from me:

Regex should be easy to remove, it's used in only one method for quick-and-dirty macro expansion. I basically only used the "\b" for word boundary detection. This means that in
```
#define VALUE 10.0
vec3 someVALUEvar = VALUE * VALUE;
```
the macro is converted into `\bVALUE\b` regex and will expand into
```
vec3 someVALUEvar = 10.0 * 10.0;
```
and NOT into
```
vec3 some10.0var = 10.0 * 10.0;
```
which will obviously cause issues. It should be pretty easy to remove the regex and replace with something else, regexp is only used in the ShaderPreprocessor::expand_macros_once(). Just make sure +-/() comma etc. and other stuff works in the replacement version if we have to drop regexp.

Also some other misc points:

* PreproprocessorTokenizer.SIZE is uppercase for no reason, can be lowercase. Some other minor code cleanup might be good too.
* Memory leak in ShaderPreprocessor::process_undef(), before removing the label from the map I forgot to memdelete() it. Easy fix.
* `#pragma disable_preprocessor` should be pretty easy to add too.
* Custom macros added from user code during runtime should be also easy to add. There are actually two "global" macros added in ever shader at the moment: OS name (not that useful) and EDITOR which it true or false if the shader is running on an editor build. Adding for example the renderer (OPENGL3/2, VULKAN) should be trivial.
* Related to that last point, preprocessing the shaders for release builds might avoid runtime compilation but it also limits things quite a bit. I think all preprocessing functionality should be available during runtime and on release builds too. This is important for example dynamic, user changeable graphical settings which might produce enormous amount of variations which are not practical (or even possible) to precompile. Or maybe even add something like `#pragma precompile` which will preprocess and compile the shader for runtime loading speed but in exchange loses the ability to change it dynamically from code by setting macros etc. But that might be a bit too overengineered?

TheOrangeDay committed:
Adding support for tracking errors from the preprocessor in the editor. Fixing memory leak issues

mhilbrunner commented:
PR meeting: we're interested in this, it needs to reviewed (by @Chaosus first, then @reduz afterwards).

Chaosus commented on editor/plugins/shader_editor_plugin.cpp:
```diff
@@ -649,6 +694,80 @@ void ShaderEditor::_make_context_menu(bool p_selection, Vector2 p_position) {
+	if (selected) {
+		String path = selected->get_metadata(0);
+
+		auto rollingCode = shader_rolling_code.find(path);
```
Don't use `auto` - it's forbidden by Juan.

akien-mga replied:
And identifiers should use `snake_case`.

Chaosus commented on servers/rendering/shader_preprocessor.cpp:
```diff
@@ -0,0 +1,955 @@
+}
+
+ShaderDependencyGraph::~ShaderDependencyGraph() {
+	for (auto E = nodes.front(); E; E = E->next()) {
```
`auto` is a forbidden keyword.

Chaosus commented on servers/rendering/shader_preprocessor.cpp:
```diff
@@ -0,0 +1,955 @@
+	shader(s) {}
+
+ShaderDependencyNode::~ShaderDependencyNode() {
+	for (auto E = dependencies.front(); E; E = E->next()) {
```
Same

Chaosus commented:
@TheOrangeDay You need to replace `auto` with an explicit declaration of the type.

TheOrangeDay commented:
@Chaosus & @akien-mga thanks for the review. I'm just in the final phase of replacing regex and hitting a few issues with terminating the preprocessor macro replacement. After that's in, I'll do the suggested updates (no file name in errors and warnings for single file edit context, adding #pragma support to disable preprocessor, and hiding the tree view by default) and then I may be ready to bring the changes to the branch this pull request is bringing in.
 
When this is done, how do I go about updating the documentation? I can also write some unit tests if I can get some guidance on that as well?

This PR has turned into a larger commitment than I orginally thought haha

Chaosus commented:
![image](https://user-images.githubusercontent.com/3036176/144460677-c7ec70c5-9ad9-4321-a2bc-ffd1d242d3e2.png)

I would also make an error emitting if #include used on the self-file or cross-reference has been detected.

TheOrangeDay committed:
Removing regex use from the preprocessor, switching auto with explicit type

TheOrangeDay committed:
Adding "View- Dependencies" to shader editor, auto hide dependency tree when there is only a single file, added inactive code color intensity to theme editor, fixed theme editing so colors update when you change them

TheOrangeDay committed:
switching remove(int) to remove_at(int) + fixing bad resolve

TheOrangeDay committed:
Merge branch 'shader-preprocessor-wip' of https://github.com/TheOrangeDay/godot into shader-preprocessor-wip

TheOrangeDay committed:
adding support for #pragma disable_preprocessor, highlighting shader dependency tree entries with a yellow background when there are warnings, and updating error and warning output to have the file path only when there are includes. Also fixed spelling mistake in name of PreprocessorTokenizer.

TheOrangeDay committed:
Improving cyclic dependency handling by tracking hash of shader code and checking included code for the same hash. There is a possibility of a collision, but there is always that chance when evaluating hashes.

TheOrangeDay committed:
updating copyright

TheOrangeDay commented:
I think this is ready for a comprehensive review now. 

I've addressed most of the comments/concerns in the PR.

- cyclic include should be detected and reported,
- added in support for #pragma disable_preprocessor
- shader deps panel doesn't show up if there are no deps
  - if there are no deps, the errors and warnings don't show a path
  - the deps panel is toggleable via "View -> Shader Dependencies"
- font color intensity of inactive code blocks due to preprocessor is adjustable in the theme editor

TheOrangeDay committed:
removing unused variable

Chaosus commented:
I think the opening #if or #ifdef requires to be closed by #endif but there is no error emitting currently:

![image](https://user-images.githubusercontent.com/3036176/148422300-27b23d68-d635-4362-b3ce-711dd8c6a982.png)

Chaosus commented:
I think you've broken auto-completion: ![bug](https://user-images.githubusercontent.com/3036176/148427021-056595c2-ffac-4406-9dea-efaaf2e4bb07.gif) Please check and fix it.

Calinou commented:
Should built-in material shaders use `#pragma disable_preprocessor` to speed up their compilation?

Ivorforce commented:
Quick comment on performance If / else uniform statements are usually optimized away at compile time. In fact, last time I checked, it was impossible to do 'true' branch statement logic in shaders - even if you can write it, it compiles away to a deterministic logic. 
I like ifdef for other reasons sometimes, but I'd really encourage checking performance in-depth before making conclusions.

Chaosus commented:
@TheOrangeDay Do you still want to continue this PR? It's in a pretty good stage - if you would fix the concerns I mentioned, I will approve it.

TheOrangeDay commented:
> @TheOrangeDay Do you still want to continue this PR? It's in a pretty good state - if you would fix the concerns I mentioned and rebase, I will approve it.

@Chaosus Sorry for the delay. I'll be looking to address those last few concerns soon.

TheOrangeDay committed:
Adding error for missing endif, adding error for invalid pragma

TheOrangeDay committed:
Fixing auto complete by changing CharType typedef in preprocessor from char to char32_t. This is to maintain the cursor char (0xFFFF).

TheOrangeDay commented:
@Chaosus I've addressed the issues you've pointed out. Let me know if you find anything else.

TheOrangeDay committed:
Removing char less than 0 check as it won't be possible to have less than zero with char32_t type. + fixing style check issues.

Chaosus commented:
Please squash the commits. Seems fine to merge now (I would improve and fix some error messages and add RTR but this may be added afterward).

bitsawer commented:
I have also been giving some quick tests and things seem to work nicely. I would fix one minor thing if possible, it might be a single line fix.

Currently when checking if a shader file has been included previously, it does not normalize paths. So for example:

```
#include "second.gdshader"
#include "NONE/../second.gdshader"
```

Includes the same file (second.gdshader) twice, even though the paths point to the same file and the second include should be skipped. Same thing can happen when including things from parent/subdirectories etc. when writing more complex shaders. Simply converting those paths always to absolute form (like "res://second.gdshader") should be fine, I think there was some simple method call to do that, but can't remember where at the moment. I think (?) the relevant code was this in "shader_preprocessor.cpp":

```
//Mark as included
state->includes.insert(real_path);
```

Another thing is that the shader code editor window is *very* narrow by default and has to be manually widened. The dependencies window takes all the space at first. Screenshot:

![godot](https://user-images.githubusercontent.com/22456603/150687994-21f2f986-3445-4303-a38f-4f67d0f94ce8.png)

GlitchedCode commented:
I've been testing this on my own, great stuff! I propose a final addition to this PR could be the ability to define macros programmatically and/or via the editor, for example by adding a field to the Shader resource that triggers a recompile, or an additional method (or overload to the `set_code()` method, to avoid needless recompilations) that takes a list of preprocessor defines and recompiles the shader accordingly.

clayjohn commented:
This looks great so far. 

One form of error checking that needs to be done is including one type of shader from another. 

For example, if a spatial shader ``#include``s a ``canvas_item`` shader, the user should get an error message that says the shader is attempting to include an incompatible type. Otherwise, when trying to import a ``canvas_item`` shader in a spatial shader (for example) you end up with a cryptic error about the usage of built in variables. 

In the past, @fire and I have discussed adding a new shader type that has not main function or built in variables, but can be #included by any other shader type. It could be called ``shader_type include`` or ``shader_type library`` or something similar. 

Another thing I noticed, but may be by design is that defines from dependencies don't carry over to parent shaders. For example, if I have a shader A
```
shader_type spatial;
#define CLAY_A 1.6
#include "B.gdshader"
#ifdef CLAY_B
// some functions
#endif
```
and a shader B
```
shader_type spatial;
#define CLAY_B 1.6
#ifdef CLAY_A
// some function
#endif
```

Shader B is able to see the defines from Shader A, and the method definitions above the #include, and Shader A can see the method definitions from Shader B, but not the defines from Shader B. In my mind (and explain to me if this is totally wrong) I expect that all the contents of Shader B should be visible in Shader A after the #include. While Shader B should have access to all the contents of Shader A defined above the #include. 

Again, great work so far! This is shaping up to be really well done.

bitsawer commented:
@clayjohn 

>For example, if a spatial shader #includes a canvas_item shader, the user should get an error message that says the shader is attempting to include an incompatible type.

I think that would limit the flexibility of include system a bit. There is a good chance that someone will want to define some constants etc. in a shader file that will be used in both spatial and canvas shaders (theming colors, animation time scale etc.). Currently the included shader_type is ignored completely and stripped away and I think that will allow users most flexibility. This means that shaders can be included or used standalone in a very flexible way (and you can even create a single shader that works BOTH ways depending on situation and defines).

>Another thing I noticed, but may be by design is that defines from dependencies don't carry over to parent shaders. 

That example seems to work for me, I can for example use CLAY_B in shader A. But maybe I misunderstood something? 
```
//In shader A, gets CLAY_B from included file.
void fragment()
{
    ALBEDO = vec3(0.0, CLAY_B, 1.0);
}
```

However, I did notice that editing dependency shader (in this case B) no longer automatically updates the assigned main shader (A) in editor if it is assigned to sprite etc. UNTIL I select the A shader in the code editor window, even if both are saved to disk.

Also, another minor parser bug(?). If an included file does not have an empty last line and ends with a define, something breaks. Simply adding this line: `#define SOMETHING 5.0` as a last line in a shader file, including it in an other file and trying to use SOMETHING breaks the shader. Simply adding a newline at the end fixes it.

Paulb23 commented on editor/code_editor.h:
```diff
@@ -244,13 +244,14 @@ class CodeTextEditor : public VBoxContainer {
 	void update_editor_settings();
 	void set_error(const String &p_error);
 	void set_error_pos(int p_line, int p_column);
+	void get_error_pos(int &p_line, int &p_column);
```
Can use a `Pointi` here rather then return args.

Chaosus replied:
Yeah, and `p_line` is an incorrect name, if the variable is intended to be returned it should be named as `r_line`.

clayjohn commented:
> @clayjohn
> 
> > For example, if a spatial shader #includes a canvas_item shader, the user should get an error message that says the shader is attempting to include an incompatible type.
> 
> I think that would limit the flexibility of include system a bit. There is a good chance that someone will want to define some constants etc. in a shader file that will be used in both spatial and canvas shaders (theming colors, animation time scale etc.). Currently the included shader_type is ignored completely and stripped away and I think that will allow users most flexibility. This means that shaders can be included or used standalone in a very flexible way (and you can even create a single shader that works BOTH ways depending on situation and defines).

I take your point. But having that level of flexibility conflicts with having useful error messages. When you have the canvas_item shader open, the error checking will be for canvas_items, so it will error if you use a built in only available in spatials (like ``ALBEDO``). This will result in weird usability issues, like being able to include a fragment function from another shader, but not being able to write to any of the outputs. In practice, users won't be able to intermix different shader types because if they do anything other than trivial examples they will run into conflicts with the names of built ins. 

An alternative solution is just to not import "main" functions from other shaders. But I have a feeling that people would like that restriction even less. I guess, it could check if the shader types differ and then not import the main function if the types are different. For example, if I include "sky.glsl" from a spatial shader, the preprocessor could know not to import the ``sky()`` function from "sky.glsl".

> > Another thing I noticed, but may be by design is that defines from dependencies don't carry over to parent shaders.
> 
> That example seems to work for me, I can for example use CLAY_B in shader A. But maybe I misunderstood something?
> 
> ```
> //In shader A, gets CLAY_B from included file.
> void fragment()
> {
>     ALBEDO = vec3(0.0, CLAY_B, 1.0);
> }
> ```

I will have to check again. When I did this I kept getting an error message saying that CLAY_B couldn't be found.

GlitchedCode commented:
After some testing i found another issue: using the preprocessor before the `shader_type` statement messes with `_get_first_ident_pos`, which in turn causes `ShaderLanguage::get_shader_type` to return bad data, which forces callers to treat shaders as spatial by default. The following screenshot demonstrates the issue:
![Screenshot_20220131_025952](https://user-images.githubusercontent.com/13611953/151729230-386a7cff-0464-4a10-994a-649ca567f0c1.png)
To fix this, `_get_first_ident_pos` in `shader_language.cpp` should be modified to also ignore, at the very least, lines that start with #.

GlitchedCode commented:
I found that using the preprocessor before the `shader_type` statement breaks the `_get_first_ident_pos` function in `shader_language.cpp`, which returns bad data and breaks callers of `ShaderLanguage::get_shader_type`. The following shader code generates an error, since it's meant to be a `canvas_item` shader but is treated as a `spatial` one:
```
#define TESTVAL 0.5
shader_type canvas_item;

#ifndef TESTVAL
	#define TESTVAL 0.0
#endif

void fragment () {
	COLOR = vec4(TESTVAL, 0.0, 0.0, 1.0); // ERROR! COLOR is a constant in fragment() of spatial shaders!
}
```
To fix this, the `_get_first_ident_pos` function should be modified to ignore, at the very least, lines starting with #, like so:
```
static int _get_first_ident_pos(const String &p_code) {
	int idx = 0;

#define GETCHAR(m_idx) (((idx + m_idx) < p_code.length()) ? p_code[idx + m_idx] : CharType(0))

	while (true) {
		// ignore comments ....
		else if (GETCHAR(0) == '#') {
			idx += 1;
			while (true)
			{
				if(GETCHAR(0) == 0) {
					return 0;
				}
				if(GETCHAR(0) == '\n'){
					idx++;
					break; // loop
				}
				idx++;
			}
		} else {
	        // ... ignore invisible characters
}
```

Chaosus commented:
@TheOrangeDay Would you like to continue to improve & rebase this PR until it's good to merge? If not, I could try to re-create it and add you as co-author.

Chaosus commented:
Supersedes by https://github.com/godotengine/godot/pull/61462.
